File: zinc-flow-java/src/main/java/zincflow/processors/TransformRecord.java
package zincflow.processors;

import org.apache.avro.Schema;
import org.apache.commons.jexl3.JexlBuilder;
import org.apache.commons.jexl3.JexlEngine;
import org.apache.commons.jexl3.JexlException;
import org.apache.commons.jexl3.JexlExpression;
import org.apache.commons.jexl3.MapContext;
import zincflow.core.FlowFile;
import zincflow.core.Processor;
import zincflow.core.ProcessorResult;
import zincflow.core.RecordContent;
import zincflow.core.SchemaDefs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringJoiner;

/// Applies a semicolon-delimited operations palette to every record in a
/// RecordContent payload. Mirrors zinc-flow-csharp's
/// {@code TransformRecord} (StdLib/ExpressionProcessors.cs:177-333).
///
/// Operations:
/// <ul>
///   <li>{@code rename:oldName:newName}     — rename a field
///   <li>{@code remove:fieldName}           — remove a field
///   <li>{@code add:fieldName:value}        — add a string-typed field with literal value
///   <li>{@code copy:source:target}         — copy field value (preserves type)
///   <li>{@code toUpper:fieldName}          — uppercase a string field
///   <li>{@code toLower:fieldName}          — lowercase a string field
///   <li>{@code default:fieldName:value}    — set field to string value if missing/null
///   <li>{@code compute:targetField:expr}   — evaluate JEXL expression, assign result
/// </ul>
///
/// {@code compute} uses Apache Commons JEXL (arithmetic, booleans, ternary,
/// string ops); C#'s track currently uses a lighter template DSL and is
/// tracked to catch up in the post-Java cohort. Every other op is bit-for-bit
/// portable across tracks.
public final class TransformRecord implements Processor {

    private static final JexlEngine JEXL = new JexlBuilder()
            .strict(false).safe(true).silent(false).create();

    private static final Set<String> KNOWN_OPS = Set.copyOf(Arrays.asList(
            "rename", "remove", "add", "copy", "toUpper", "toLower", "default", "compute"));

    private record Directive(String op, String arg1, String arg2, JexlExpression compiled) {}

    private final List<Directive> directives;

    public TransformRecord(String operationsSpec) {
        if (operationsSpec == null || operationsSpec.isBlank()) {
            throw new IllegalArgumentException(
                    "TransformRecord: operations must have at least one directive");
        }
        List<Directive> parsed = new ArrayList<>();
        for (String raw : operationsSpec.split(";")) {
            String directive = raw.trim();
            if (directive.isEmpty()) continue;
            // Split into up to 3 parts so arg2 can legally contain colons
            // (e.g. compute expressions with dotted paths).
            String[] parts = directive.split(":", 3);
            if (parts.length < 2) {
                throw new IllegalArgumentException(
                        "TransformRecord: malformed directive '" + directive
                                + "' — expected 'op:arg1' or 'op:arg1:arg2'");
            }
            String op = parts[0];
            if (!KNOWN_OPS.contains(op)) {
                throw new IllegalArgumentException(
                        "TransformRecord: unknown op '" + op + "' in directive '" + directive
                                + "' — valid: " + String.join(", ", KNOWN_OPS));
            }
            String arg1 = parts[1];
            String arg2 = parts.length > 2 ? parts[2] : "";
            JexlExpression compiled = null;
            if ("compute".equals(op)) {
                if (arg2.isBlank()) {
                    throw new IllegalArgumentException(
                            "TransformRecord: compute requires an expression — 'compute:target:expr'");
                }
                try {
                    compiled = JEXL.createExpression(arg2);
                } catch (JexlException ex) {
                    throw new IllegalArgumentException(
                            "TransformRecord: invalid JEXL in compute '" + arg2 + "' — " + ex.getMessage(), ex);
                }
            }
            parsed.add(new Directive(op, arg1, arg2, compiled));
        }
        if (parsed.isEmpty()) {
            throw new IllegalArgumentException("TransformRecord: operations spec produced no directives");
        }
        this.directives = List.copyOf(parsed);
    }

    @Override
    public ProcessorResult process(FlowFile ff) {
        if (!(ff.content() instanceof RecordContent rc)) {
            return ProcessorResult.failure(
                    "TransformRecord: expected RecordContent, got " + ff.content().getClass().getSimpleName(), ff);
        }
        if (rc.records().isEmpty()) {
            return ProcessorResult.single(ff);
        }

        List<Map<String, Object>> transformed = new ArrayList<>(rc.records().size());
        for (Map<String, Object> record : rc.records()) {
            Map<String, Object> dict = new LinkedHashMap<>(record);
            for (Directive d : directives) {
                applyOp(dict, d);
            }
            transformed.add(dict);
        }

        // Re-infer schema from first record post-transform. Unlike C# we
        // don't track per-field original Avro types — the safest choice
        // is a fresh inferred schema, which covers the common cases
        // (add/compute adds typed fields; rename/copy preserves types
        // through the map; remove drops fields cleanly).
        String schemaName = rc.schema() == null ? "TransformedRecord" : rc.schema().getName();
        Schema outSchema = inferSchema(schemaName, transformed.getFirst());

        return ProcessorResult.single(ff.withContent(new RecordContent(transformed, outSchema)));
    }

    private void applyOp(Map<String, Object> dict, Directive d) {
        switch (d.op()) {
            case "rename" -> {
                if (dict.containsKey(d.arg1())) dict.put(d.arg2(), dict.remove(d.arg1()));
            }
            case "remove" -> dict.remove(d.arg1());
            case "add"    -> dict.put(d.arg1(), d.arg2());
            case "copy"   -> {
                if (dict.containsKey(d.arg1())) dict.put(d.arg2(), dict.get(d.arg1()));
            }
            case "toUpper" -> {
                if (dict.get(d.arg1()) instanceof String s) dict.put(d.arg1(), s.toUpperCase());
            }
            case "toLower" -> {
                if (dict.get(d.arg1()) instanceof String s) dict.put(d.arg1(), s.toLowerCase());
            }
            case "default" -> {
                if (!dict.containsKey(d.arg1()) || dict.get(d.arg1()) == null) dict.put(d.arg1(), d.arg2());
            }
            case "compute" -> {
                MapContext ctx = new MapContext();
                ctx.set("record", dict);
                for (var e : dict.entrySet()) ctx.set(e.getKey(), e.getValue());
                try {
                    dict.put(d.arg1(), d.compiled().evaluate(ctx));
                } catch (JexlException ex) {
                    // C# silently skips failed computes — match that so
                    // one bad record doesn't crash the whole batch.
                }
            }
            default -> { /* unreachable — parse-time validation rejects unknown ops */ }
        }
    }

    /// Build a flat Avro Schema from a record's keys + value types.
    /// Matches the inference logic in ConvertJSONToRecord / ConvertCSVToRecord.
    private static Schema inferSchema(String name, Map<String, Object> record) {
        if (record.isEmpty()) return null;
        StringJoiner defs = new StringJoiner(",");
        Set<String> seen = new HashSet<>();
        for (var entry : record.entrySet()) {
            if (!seen.add(entry.getKey())) continue;
            defs.add(entry.getKey() + ":" + inferType(entry.getValue()));
        }
        return SchemaDefs.parse(name, defs.toString());
    }

    private static String inferType(Object v) {
        if (v == null) return "string";
        if (v instanceof Boolean) return "boolean";
        if (v instanceof Integer) return "int";
        if (v instanceof Long) return "long";
        if (v instanceof Float) return "float";
        if (v instanceof Double) return "double";
        if (v instanceof byte[]) return "bytes";
        return "string";
    }
}

File: caravan-out/processors/transform_record.go
package processors

import (
	"fmt"
	"zinc-flow/core"
)

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:30
type RecordOp struct {
	Op string
	Arg1 string
	Arg2 string
}

func NewRecordOp(op string, arg1 string, arg2 string) RecordOp {
	return RecordOp{Op: op, Arg1: arg1, Arg2: arg2}
}

func (s RecordOp) String() string {
	return fmt.Sprintf("RecordOp(op=%v, arg1=%v, arg2=%v)", s.Op, s.Arg1, s.Arg2)
}

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:32
type TransformRecord struct {
	operations []RecordOp
}

func NewTransformRecord(operationsSpec string) *TransformRecord {
	s := &TransformRecord{operations: []RecordOp{}}
	s.parseOperations(operationsSpec)
	return s
}

func (s *TransformRecord) parseOperations(spec string) {
	if spec == "" {
		return
	}
	start := 0
	i := 0
	for i < len(spec) {
		if spec[i:i + 1] == ";" {
			s.addEntry(spec[start:i])
			start = i + 1
		}
		i = i + 1
	}
	if start < len(spec) {
		s.addEntry(spec[start:len(spec)])
	}
}

func (s *TransformRecord) addEntry(entry string) {
	trimmed := trimOpSpace(entry)
	if trimmed == "" {
		return
	}
	parts := []string{}
	partStart := 0
	maxParts := 3
	j := 0
	for j < len(trimmed) && len(parts) < maxParts - 1 {
		if trimmed[j:j + 1] == ":" {
			parts = append(parts, trimOpSpace(trimmed[partStart:j]))
			partStart = j + 1
		}
		j = j + 1
	}
	parts = append(parts, trimOpSpace(trimmed[partStart:len(trimmed)]))
	if len(parts) == 0 || parts[0] == "" {
		return
	}
	arg1 := ""
	arg2 := ""
	if len(parts) >= 2 {
		arg1 = parts[1]
	}
	if len(parts) >= 3 {
		arg2 = parts[2]
	}
	s.operations = append(s.operations, NewRecordOp(parts[0], arg1, arg2))
}

func (s *TransformRecord) Process(ff core.FlowFile) core.ProcessorResult {
	switch _v := ff.Content.(type) {
	case core.Records:
		schema := _v.Schema
		records := _v.Rows
		if len(records) == 0 {
			return core.NewSingle(ff)
		}
		transformed := []*core.GenericRecord{}
		for _, record := range records {
			transformed = append(transformed, s.applyOps(record, schema))
		}
		newSchema := rebuildSchema(schema, transformed[0])
		updated := core.WithContent(ff, core.NewRecords(newSchema, transformed))
		return core.NewSingle(updated)
	default:
		_ = _v
		return core.NewSingle(ff)
	}
	return nil
}

func (s *TransformRecord) applyOps(record *core.GenericRecord, schema core.Schema) *core.GenericRecord {
	values := map[string]interface{}{}
	for _, f := range schema.Fields {
		v := record.GetField(f.Name)
		if v != nil {
			values[f.Name] = v
		}
	}
	for _, op := range s.operations {
		s.applyOne(values, op)
	}
	fields := []core.Field{}
	origByName := map[string]core.FieldType{}
	for _, f := range schema.Fields {
		origByName[f.Name] = f.FieldType
	}
	for _, k := range func() []string { _keys := make([]string, 0, len(values)); for _k := range values { _keys = append(_keys, _k) }; return _keys }() {
		ft := core.STRING
		if func() bool { _, _ok := origByName[k]; return _ok }() {
			ft = origByName[k]
		}
		fields = append(fields, core.NewField(k, ft, nil))
	}
	outSchema := core.NewSchema(schema.Name, fields)
	out := core.NewGenericRecord(outSchema)
	for _, k := range func() []string { _keys := make([]string, 0, len(values)); for _k := range values { _keys = append(_keys, _k) }; return _keys }() {
		out.SetField(k, values[k])
	}
	return out
}

func (s *TransformRecord) applyOne(values map[string]interface{}, op RecordOp) {
	if op.Op == "rename" {
		if func() bool { _, _ok := values[op.Arg1]; return _ok }() {
			values[op.Arg2] = values[op.Arg1]
			delete(values, op.Arg1)
		}
		return
	}
	if op.Op == "remove" {
		delete(values, op.Arg1)
		return
	}
	if op.Op == "add" {
		values[op.Arg1] = op.Arg2
		return
	}
	if op.Op == "copy" {
		if func() bool { _, _ok := values[op.Arg1]; return _ok }() {
			values[op.Arg2] = values[op.Arg1]
		}
		return
	}
	if op.Op == "toUpper" {
		if func() bool { _, _ok := values[op.Arg1]; return _ok }() {
			v := values[op.Arg1]
			if v != nil {
				values[op.Arg1] = asciiUpper(fmt.Sprint(v))
			}
		}
		return
	}
	if op.Op == "toLower" {
		if func() bool { _, _ok := values[op.Arg1]; return _ok }() {
			v := values[op.Arg1]
			if v != nil {
				values[op.Arg1] = asciiLower(fmt.Sprint(v))
			}
		}
		return
	}
	if op.Op == "default" {
		if !func() bool { _, _ok := values[op.Arg1]; return _ok }() || values[op.Arg1] == nil {
			values[op.Arg1] = op.Arg2
		}
		return
	}
}


//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:203
func TransformRecordFactory(ctx *core.ScopedContext, config map[string]string) core.ProcessorFn {
	spec := ""
	if func() bool { _, _ok := config["operations"]; return _ok }() {
		spec = config["operations"]
	}
	return NewTransformRecord(spec)
}

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:215
func rebuildSchema(orig core.Schema, first *core.GenericRecord) core.Schema {
	origFields := map[string]core.FieldType{}
	for _, f := range orig.Fields {
		origFields[f.Name] = f.FieldType
	}
	firstSchema := first.GetSchema()
	newFields := []core.Field{}
	for _, f := range firstSchema.Fields {
		ft := f.FieldType
		if func() bool { _, _ok := origFields[f.Name]; return _ok }() {
			ft = origFields[f.Name]
		}
		newFields = append(newFields, core.NewField(f.Name, ft, nil))
	}
	return core.NewSchema(orig.Name, newFields)
}

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:232
func trimOpSpace(s string) string {
	start := 0
	for start < len(s) {
		c := s[start:start + 1]
		if c != " " && c != "\t" {
			break
		}
		start = start + 1
	}
	stop := len(s)
	for stop > start {
		c := s[stop - 1:stop]
		if c != " " && c != "\t" {
			break
		}
		stop = stop - 1
	}
	return s[start:stop]
}

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:252
func asciiUpper(s string) string {
	out := ""
	i := 0
	for i < len(s) {
		c := s[i:i + 1]
		if c >= "a" && c <= "z" {
			b := byte(s[i])
			out = out + string(byte(b - 32))
		} else {
			out = out + c
		}
		i = i + 1
	}
	return out
}

//line /home/vrjoshi/proj/zinc-flow/src/processors/transform_record.zn:268
func asciiLower(s string) string {
	out := ""
	i := 0
	for i < len(s) {
		c := s[i:i + 1]
		if c >= "A" && c <= "Z" {
			b := byte(s[i])
			out = out + string(byte(b + 32))
		} else {
			out = out + c
		}
		i = i + 1
	}
	return out
}

